plugins/utils/which-key.nix

```diff
@@ -276,36 +521,28 @@ with lib;
             padding = mkSpacing cfg.window.padding;
             inherit (cfg.window) winblend;
           };
-        inherit (cfg) layout;
-        ignore_missing = cfg.ignoreMissing;
-        inherit (cfg) hidden;
-        show_help = cfg.showHelp;
-        show_keys = cfg.showKeys;
-        inherit (cfg.settings) spec;
-        inherit (cfg) triggers;
-        triggers_nowait = cfg.triggersNoWait;
-        triggers_blacklist = cfg.triggersBlackList;
-        inherit (cfg) disable;
       };
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-
-      warnings = optional opt.registrations.isDefined ''
-        nixvim (plugins.which-key):
-        The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
-        You should use `plugins.which-key.settings.spec' instead.
 
-        Note: the spec format has changed in which-key v3
-        See: https://github.com/folke/which-key.nvim?tab=readme-ov-file#%EF%B8%8F-mappings
-      '';
+      # TODO: introduced 2024-08-05: remove after 24.11
+      warnings =
+        optional opt.registrations.isDefined ''
+          nixvim (plugins.which-key):
+          The option definition `plugins.which-key.registrations' in ${showFiles opt.registrations.files} has been deprecated in which-key v3; please remove it.
+          You should use `plugins.which-key.settings.spec' instead.
 
-      extraConfigLua =
-        ''
-          require("which-key").setup(${helpers.toLuaObject setupOptions})
+          Note: the spec format has changed in which-key v3
+          See: https://github.com/folke/which-key.nvim?tab=readme-ov-file#%EF%B8%8F-mappings
         ''
-        + (optionalString (opt.registrations.isDefined && cfg.registrations != { }) ''
-          require("which-key").register(${helpers.toLuaObject cfg.registrations})
-        '');
+        ++ optional windowIsDefined ''
+          nixvim (plugins.which-key):
+          The option definition `plugins.which-key.window' in ${showFiles opt.window.border.files} has been deprecated in which-key v3; please remove it.
+          You should use `plugins.which-key.settings.win' instead.
+
+          Note: the window options have changed parameters in which-key v3
+        '';
+
+      extraConfigLua = optionalString (opt.registrations.isDefined && cfg.registrations != { }) ''
+        require("which-key").register(${toLuaObject cfg.registrations})
+      '';
```

Comment:
Now that the only line being set `extraConfigLua` is the optional one, we should probably not use `optionalString`.

When the condition is false, an empty string will be defined which will result in a blank line being needlessly added to init.lua.

Normally, we'd use `mkIf` here, however that's won't do what we want; `cfg.registrations` would still be evaluated even when `opt.registrations.isDefined = false`.

The "correct" solution would be to either use `mkMerge` or an `//` update combined with `optionalAttrs`, e.g:

```nix
extraConfig = cfg: mkMerge [
  {/* main config */}
  (optionalAttrs opt.registrations.isDefined {
    # config that uses cfg.registrations
  })
];
```

Reply:
See #1909

Reply:
Not seeing an empty string in my generated config with this and when I tried using the mkMerge on extraConfig I kept getting infinite recursion.